### PR TITLE
Allow to create static OIDC tenants programmatically

### LINKED ISSUE #44958: Programmatic OIDC startup
### Description

Currently, Quarkus OIDC users usually use properties to initiate the OIDC startup (connection, discovery, JWK set retrieval).
They can also use `TenantConfigResolver` to create OIDC tenants dynamically - but it happens at the request time.

It would be good to support programmatic OIDC startup, for users to be able to do it in `main()` or in Startup event handlers.

This enhancement will serve 2 purposes:

* Help users who are not too keen on dealing with properties just to do some simple code and be done with it.
* Help users who are migrating to Quarkus OIDC and prefer the programmatic setup

### Implementation ideas

Now that we have an OIDC builder support, we can try something like:
```
public void main() {
    // Enough for simple bearer token scenarios
    Oidc.start(OidcTenantConfigBuilder.authServerUrl("....").build());

   // code flow
   Oidc.start(OidcTenantConfigBuilder.authServerUrl("....").applicationType(WEB_APP)
        clientId(...)
        credentials().clientSecret("").end()
  .build());
    
}
``` 

etc

### PR DESCRIPTION
**Pull request by michalvavrik**
- closes: https://github.com/quarkusio/quarkus/issues/44958
- allows to programmatically configure static OIDC tenants; there are features restricted to static tenants (like issuer-based and path-based tenant resolvers) can use tenants created like `OidcTenantConfig.authServerUrl("...").build()`

### CODE REVIEW

File: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc [hunk 1]
```diff
@@ -1464,6 +1464,37 @@ public class DiscoveryEndpointResponseFilter implements OidcResponseFilter {
...
+package io.quarkus.it.oidc;
+
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) {
+        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/quarkus").build());
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc:1482]: Proposing a shortcut, `Oidc.create("http://localhost:8180/realms/quarkus")`.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 2]
```diff
@@ -2049,6 +2049,47 @@ quarkus.log.category."io.quarkus.oidc.runtime.OidcRecorder".min-level=TRACE
...
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+import static io.quarkus.oidc.runtime.OidcTenantConfig.ApplicationType.WEB_APP;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) {
+        oidc.create(OidcTenantConfig
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:2069]: Proposing a shortcut: `Oidc.createWebApp("http://localhost:8180/realms/quarkus", "quarkus-app", "my-secret")`

File: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc (hunk 1 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc:1482]: This has 2 issues:

- we already have `create` method that is not static, so we can't have one `create` method static and one non-static, both with same signature:
  - if you proposed different name than of course it is trivial to do, we just do inside that static `create` method something like `Arc.container().select(Oidc.class).createWhatver(...)`
- timing is key, using `Oidc.create` must be primarily done before OIDC has started, which is why I have chosen the observer method; I can elaborate why is it so important if it is not clear, just let me know

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc:1482]: ah, I get it. you don't mind static, you suggest that `oidc.create(`auth-server-url`)` simply builds the tenant right? I misunderstood you

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 3]
```diff
@@ -1110,6 +1110,111 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+package io.quarkus.it.oidc;
+
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc, OidcConfig config) {
+        oidc.create(OidcTenantConfig
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1167]: Looks interesting

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 4]
```diff
@@ -1110,6 +1110,111 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+
+    private String getAdditionalTenantPath() {
+        return "/additional-tenant-path";
+    }
+
+}
+----
+<1> The tenant `tenant-two` will have two paths, `/additional-tenant-path` and `/api/tenant-two/*`.
+
+Dynamic OIDC tenants resolved with the <<annotations-tenant-resolver>> can be created whenever necessary:
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1182]: I'm not sure about this one. Annotations can only be used to resolving static tenants AFAIK ?

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 5]
```diff
@@ -1110,6 +1110,111 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+----
+package io.quarkus.it.oidc;
+
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) { <1>
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: Should it really be an event as well ? IMHO having it injected is better, and used in on start up events. It may be a bit confusing otherwise.

By the way, how does using `Oidc` work in `main()` ?

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 4 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1182]: I changed that :-) I have a test for it. It is not important change, I can drop it. But in that case, I don't see a point of keeping `Oidc` as a bean. We could drop it in that case.

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 5 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: The question is probably the most important part - it is really important that we don't pledge ourselves to start Oidc after all the startup events has been fired, because we don't know what else are users doing during the startup that may require Oidc. If we agree that users can create static tenants with the startup event, we also agree to start Oidc after all the startup events has been processed (sync or async) but we need to do it before HTTP server has started. After all the startup events has been fired, only thing that I know that happens is starting the HTTP server (sure, I could be wrong) and we will need to get somewhere in between. I think it is better to own our own timing.

I'll throw few arguments because this is topic I have thought about a lot. Please let me know if they don't make sense:

1. We fire this when bean container is completely ready.
2. We fire this when runtime config is completely ready.
3. We need `TenantConfigBean` ready before HTTP router is finalized (no the same as HTTP server started) before than, even with `Router` is fired and `io.quarkus.oidc.runtime.BackChannelLogoutHandler` needs to use it.  BTW why shouldn't we send `Oidc` as event when other events like `Router` are acceptable?

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: > By the way, how does using Oidc work in main() ?

Do you mean this https://quarkus.io/guides/lifecycle#the-main-method? Docs says `It is not recommenced to do any business logic in this main method, as Quarkus has not been set up yet` so no, it won't and shouldn't.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: > IMHO having it injected is better

We need to differ between static and dynamic tenants. The static ones are final and we must have them before Oidc starts. You can inject `Oidc` and create dynamic tenants.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: Not sure if it is clear why I think this is so important. And I can be wrong...  Just for case - some features are only available for the static tenants and they are setup once (like issue/path-based resolvers, backchannel logout etc.) and we need to have static tenants ready when we do that. If we allow users to create static tenants anytime, we will have to refactor a lot of things and I think it will give only little fruit. We need static tenants ready before users use OIDC and if we own logic when we fire the event with the `OIDC`, then we have flexibility. We can always postpone it when necessary as long as the HTTP server is not running. But once we say: _you can create static tenants in the startup events_, we will never be able to move if before that without potentially breaking changes.

File: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc (hunk 1 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc:1482]: Hi @michalvavrik, sure, that was just the first review try.

Yes, sorry, in the `@Observes` case it is `oidc.create(url)`;

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 2 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:2069]: Oops, same here, lower case `oidc` :-)

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 4 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1182]: Hi Michal, can we deal with it later ? I'm not really sure about talking about static and dynamic tenants in scope of the initial PR introducing this feature. It would be definitely a dedicated PR IMHO in any case. 

FYI, it sounds promising, the whole idea of allocating tenants dynamically to match the annotations. The dynamism is limited but it kind of introduces some tenant-name and request path restriction, but please, let's not deal with it in this PR

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 5 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: Hi Michal, @michalvavrik, so `Oidc` is the startup kind of event, right ? (And sure, if it also works for `Router` then it only makes sense to allow it for `Oidc`).

I'd only ask for us not to go into the dynamic OIDC tenants space at this stage.

As far as I understand, the main reason some users prefer the programmatic approach is because they don't want properties. And we already let users create dynamic tenants with `TenantConigResolver`.

So, it appears to me the main use case, right now at least, is to create, possibly a single (static) tenant programmatically at startup.

We can definitely evolve and grow it further, but I'd start with the simplest possible option

Thanks

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: Cool @sberyozkin ; I also think the main use case are static tenants. I can propose allowing registering dynamic tenants  later together with the `@Tenant` annotation. Static tenants were my main motivation. I'll put to it to my list and propose something in regards of dynamic tenants in the second half of January. It should make this PR little bit easier to review.

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 4 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1182]: I'll drop registering dynamic tenants from this PR and we can discuss it in a separate PR later. I don't think it is important use case

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 5 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1128]: > Hi Michal, @michalvavrik, so Oidc is the startup kind of event, right ?

Yes, I suspect it will happen right before the startup event, but I don't have working knowledge of build steps ordering, there is some asynchronous mechanism I have never checked.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package io.quarkus.oidc;
+
+import io.quarkus.oidc.common.runtime.config.OidcClientCommonConfig;
+import io.quarkus.oidc.runtime.OidcTenantConfig.ApplicationType;
+
+/**
+ * A CDI event that facilitates programmatic OIDC extension setup.
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:7]:
```suggestion
 * A CDI event that facilitates programmatic OIDC provider setup.
```

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * Static OIDC tenants must be created in a synchronous manner from an observer method like in the example below:
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:8]: I'd like to propose not to highlight a rather technical `static tenant` term, so the follow up suggestions are all related to it:

```suggestion
 * OIDC providers must be created from an observer method like in the example below:
```

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * <pre>
+ * {@code
+ * public class OidcSetup {
+ *
+ *     void createStaticTenant(@Observes Oidc oidc) {
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:14]:
```suggestion
 *     void createDefaultTenant(@Observes Oidc oidc) {
```

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * }
+ * </pre>
+ *
+ * The example above is equivalent to configuring {@code quarkus.oidc.auth-server-url=https://oidc-provider-hostname/}
+ * in the application.properties.
+ */
+public interface Oidc {
+
+    /**
+     * Creates static OIDC tenant. This method can also be used to create static tenants based on the tenants
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:28]:
```suggestion
     * Creates OIDC tenant.
```

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+     * configured in the 'application.properties' file or any other configuration source.
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:29]:
```suggestion
```

I'll comment further on it in the multi-tenancy doc update

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param tenantConfig tenant config; must not be null
+     */
+    void create(OidcTenantConfig tenantConfig);
+
+    /**
+     * Creates default static OIDC tenant with the {@link ApplicationType#SERVICE} application type.
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:36]:
```suggestion
     * Creates default OIDC tenant with the {@link ApplicationType#SERVICE} application type.
```

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param authServerUrl {@link OidcTenantConfig#authServerUrl()}
+     */
+    void createServiceApp(String authServerUrl);
+
+    /**
+     * Creates default static OIDC tenant with the {@link ApplicationType#WEB_APP} application type.
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:43]:
```suggestion
     * Creates default OIDC tenant with the {@link ApplicationType#WEB_APP} application type.
```

File: docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc [hunk 13]
```diff
@@ -1464,6 +1464,36 @@ public class DiscoveryEndpointResponseFilter implements OidcResponseFilter {
...
+}
+----
+
+The code above is a programmatic equivalent to the following configuration in the `application.properties` file:
+
+[source,properties]
+----
+quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
+----
+
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-bearer-token-authentication.adoc:1493]: I propose here to show another example as you show it in Java Docs where `OidcConfig` is created, so that it is obvious users can go full OIDC config with the builder if required

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 14]
```diff
@@ -2049,6 +2049,39 @@ quarkus.log.category."io.quarkus.oidc.runtime.OidcRecorder".min-level=TRACE
...
+The code above is a programmatic equivalent to the following configuration in the `application.properties` file:
+
+[source,properties]
+----
+quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
+quarkus.oidc.application-type=web-app
+quarkus.oidc.client-id=quarkus-app
+quarkus.oidc.credentials.secret=mysecret
+----
+
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:2081]: Please show the `OidcTenantConfig` way as well

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+ */
+public interface Oidc {
+
+    /**
+     * Creates static OIDC tenant. This method can also be used to create static tenants based on the tenants
+     * configured in the 'application.properties' file or any other configuration source.
+     *
+     * @param tenantConfig tenant config; must not be null
+     */
+    void create(OidcTenantConfig tenantConfig);
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:33]: Do we have any risk of blocked threads here since it is happening on the CDI event call ?

May be we can do something like this in the future: allow accepting async events, and `createAsync` methods will mark which tenant is being created right now, and then, when the request arrives and the tenant resolver determines that the current tenant is still being created then it will wait until it is initialized ?

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java [hunk 16]
```diff
@@ -662,6 +663,9 @@ public OidcTenantConfigBuilder provider(Provider provider) {
      * @return build {@link io.quarkus.oidc.OidcTenantConfig}
      */
     public io.quarkus.oidc.OidcTenantConfig build() {
+        if (tenantId.isEmpty()) {
```
Comment (sberyozkin) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java:666]: Would it make sense to have a dedicated PR for this fix as it won't only apply to this PR ?

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 17]
```diff
@@ -1110,6 +1110,75 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
 
 Be aware that tenant-specific configurations can also be disabled, for example: `quarkus.oidc.tenant-a.tenant-enabled=false`.
 
+[[programmatic-startup]]
+== Programmatic OIDC start-up for multitenant application
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1114]:
```suggestion
== Programmatic OIDC start-up for multiple tenants
```

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 18]
```diff
@@ -1110,6 +1110,75 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) { <1>
+        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/tenant-one").tenantId("tenant-one").build());
+        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/tenant-two").tenantId("tenant-two").build());
+    }
+
+}
+----
+<1> Static tenants must be created in a synchronous manner from the observer method.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1135]:
```suggestion
<1> Observe OIDC event.
```

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 19]
```diff
@@ -1110,6 +1110,75 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) { <1>
+        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/tenant-one").tenantId("tenant-one").build());
+        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/tenant-two").tenantId("tenant-two").build());
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1130]:
```suggestion
        oidc.create(OidcTenantConfig.authServerUrl("http://localhost:8180/realms/tenant-two").tenantId("tenant-two").build()); <3>
```
And add below `<> Create OIDC tenant 'tenant-two'.`

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc [hunk 20]
```diff
@@ -1110,6 +1110,75 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
...
+
+The code above is a programmatic equivalent to the following configuration in the `application.properties` file:
+
+[source,properties]
+----
+quarkus.oidc.tenant-one.auth-server-url=http://localhost:8180/realms/tenant-one
+quarkus.oidc.tenant-two.auth-server-url=http://localhost:8180/realms/tenant-two
+----
+
+You can also create static tenants based on the tenants configured in the `application.properties` file.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1145]: As I said, it does look interesting, but I was thinking about it, and started having some doubts:
* It is a fairly advanced case.
* Here we document accessing a runtime package `OidcConfig`, `namedTenants`, we make it part of the API
* This raises the whole lot of questions: what if the users change other properties available at the builder level

IMHO, it is a bit early to start documenting this option. 
Michal, honestly, there is something interesting here - where we let users customize configured tenants, thanks for keeping coming up with these new ideas, but IMHO it requires more discussions, please don't hesitate to open new issue to record your ideas (this one, and dynamic tenant allocation to match annotations)

**Review by sberyozkin — COMMENTED** [review]
Michal, IMHO, another brilliant piece of work, with a very big impact. Thanks a million, let's tune a few minor details and it will be ready to go

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java (hunk 15 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:33]: > Do we have any risk of blocked threads here since it is happening on the CDI event call ?

No, I don't think there is a risk. Remember that you also block threads when creating static tenant contexts, I don't think there is any difference here because it is almost same moment and definitely the same thread.

> May be we can do something like this in the future: allow accepting async events, and createAsync methods will mark which tenant is being created right now, and then, when the request arrives and the tenant resolver determines that the current tenant is still being created then it will wait until it is initialized ?

We definitely could do `createAsync`, then subscribe to all joined `Uni` and only create OIDC when is ready. There will need to be timeouts due, because while we can postpone creating OIDC ourselves, we cannot block others from creating it, because if there are incoming HTTP requests, it is going to be created anyway.

Async events alone won't do IMHO, because we can subscribe to the completion of the async events, but just because it is called "async", there isn't anything async about it. `@ObservesAsync` only says that other code after firing the event is not waiting till the event is processing. But it doesn't help you to for example doing async HTTP requests better than the normal event.

TL;DR; yes, I agree we can do better, but I'd probably wait till someone asks for do something about dynamic tenants. I am bit worried that if we pledge to postpone creating OIDC too much, will will start having little "bugs" or issues that this `oidc.create(...)` doesn't work for some combination with other extensions or whatever.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java (hunk 16 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfigBuilder.java:666]: `OidcTenantConfigBuilder` is only in the main and in no released version, so I think it is not necessary. (unless I misunderstood it, this change is about allowing users to create default tenants without setting default tenant id explicitly, which I strongly suspect most of the users don't know where to find).

File: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc (hunk 20 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc:1145]: The point of this example was mostly:

- show you can inject config properties because they are going to be ready (whether it is OidcConfig)
- I suspect that there will be always properties that you know before hand and doesn't change for your use case, so there is no point having them hardcoded in Java. I don't know what is mainstream, but the company I worked in past really hated hardcoding stuff in Java, so I think users would appreciate knowing they can only set _some_ dynamic properties

I'll drop this from documentation for now. I don't think we should block this PR by discussing it too much here, but IMHO we should come back to this, I have added this on my list and will come with it (either in form of issue or discussion)

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java (hunk 15 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:33]: btw I wrote in the Javadoc about "synchronous manner" but I only wanted users to create tenants immediately and not in the subscriber or on completion etc.

Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/Oidc.java:32]: @michalvavrik OK, let's consider it later

**Review by sberyozkin — APPROVED** [review]
